In [ ]:
# First steps - working environment

I have three kernels in my jupyter environment - Python3 (default), R 3.3, and BASH. 
The vast majority of this will be done in the BASH terminal, with a lot of the more complex text editing done with python scripts.

In [ ]:
## graham and I will presumably have our reads saved in different paths
## so I will save my path to the raw reads and the path to my
## working directory as a shell variable for the moment.

## change to my directory that contains the raw reads:
cd /home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls

saving a variable name for working directory. This will make our code more universal. 

In [1]:
wdir='/home/daniel/Documents/Taiwan_data/wood/woodbiom'

## read directory, holds our raw reads:
rdir='/home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls/woodonly'

In [20]:
## we can use these instead of our individualized path variables:

echo $rdir
cd $wdir
pwd

/home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls/woodonly
/home/daniel/Documents/Taiwan_data/wood/woodbiom


Generally, let's try to stay in our working directory, and bring other files as we require them

What do our read files look like?

In [42]:
ls $rdir | head -n 20

index.html
lane1-s001-index-CAACACCT-TTGCTTGGNNNN-L1_S1_L001_R1_001.fastq
lane1-s001-index-CAACACCT-TTGCTTGGNNNN-L1_S1_L001_R2_001.fastq.gz
lane1-s002-index-CAACACCT-CCTGTCAANNNN-L2_S2_L001_R1_001.fastq.gz
lane1-s002-index-CAACACCT-CCTGTCAANNNN-L2_S2_L001_R2_001.fastq.gz
lane1-s003-index-CAACACCT-AGCCTATCNNNN-L3_S3_L001_R1_001.fastq.gz
lane1-s003-index-CAACACCT-AGCCTATCNNNN-L3_S3_L001_R2_001.fastq.gz
lane1-s004-index-CAACACCT-TGATCACGNNNN-L4_S4_L001_R1_001.fastq.gz
lane1-s004-index-CAACACCT-TGATCACGNNNN-L4_S4_L001_R2_001.fastq.gz
lane1-s005-index-CTAGGTGA-TTGCTTGGNNNN-Fr13_S5_L001_R1_001.fastq.gz
lane1-s005-index-CTAGGTGA-TTGCTTGGNNNN-Fr13_S5_L001_R2_001.fastq.gz
lane1-s006-index-CTAGGTGA-CCTGTCAANNNN-Fr14_S6_L001_R1_001.fastq.gz
lane1-s006-index-CTAGGTGA-CCTGTCAANNNN-Fr14_S6_L001_R2_001.fastq.gz
lane1-s007-index-CTAGGTGA-AGCCTATCNNNN-Fr15_S7_L001_R1_001.fastq.gz
lane1-s007-index-CTAGGTGA-AGCCTATCNNNN-Fr15_S7_L001_R2_001.fastq.gz
lane1-s008-index-CTAGGTGA-TGATCACGNNNN-Fr16_S8_L001_R1_0

Lots of archived zipped files. Should be one forward (R1, I think) 
and one reverse (R2) for each of the samples submitted. 
What do these files look like unzipped? Let's check my first sample:

In [44]:
ls $rdir | grep '\-1w_'

lane1-s164-index-CTAGGTGA-GACACAGTNNNN-1w_S164_L001_R1_001.fastq.gz
lane1-s164-index-CTAGGTGA-GACACAGTNNNN-1w_S164_L001_R2_001.fastq.gz


In [47]:
## copy these two files, my first sample forward and reverse, to my working directory
cp $rdir/lane1-s164-index-CTAGGTGA-GACACAGTNNNN-1w_S164_L001_R*_001.fastq.gz $wdir

## unzip
gunzip lane1-s164-index-CTAGGTGA-GACACAGTNNNN-1w_S164_L001_R1_001.fastq.gz
gunzip lane1-s164-index-CTAGGTGA-GACACAGTNNNN-1w_S164_L001_R2_001.fastq.gz

In [54]:
## how large are these? 
wc -l lane1-s164-index-CTAGGTGA-GACACAGTNNNN-1w_S164_L001_R1_001.fastq
wc -l lane1-s164-index-CTAGGTGA-GACACAGTNNNN-1w_S164_L001_R2_001.fastq

198472 lane1-s164-index-CTAGGTGA-GACACAGTNNNN-1w_S164_L001_R1_001.fastq
198472 lane1-s164-index-CTAGGTGA-GACACAGTNNNN-1w_S164_L001_R2_001.fastq


a good sign that they match in size...
take a peek inside:

In [57]:
tail *R2*fastq

+
<@CC8FGGGGGGGGGGGGGGGGGGFG-CFFC6C6FG9C@;CC,,6,,CFF<FC+F,B:F@F6F,C6,,C+86@D@<,CF<C,,<CF8@++7@F:=>,4,:,,49C4C+++84+8+4+8:B46:++84@++3@3+,,,,@B<>,@93;,@@F,,333<,,@,,***41**,*2,,,,*416**38**3*/*37<F7+1+3=5>8:76*//)5*20+:CGG7<0,*)1*)****9<A)))(*-):**1*7*3((5)/9<((030)))-))/).*)1).))-5)))6)))**)****/*1*)33
@M01498:244:000000000-ANT97:1:2119:12315:25243 2:N:0:164
CAGAACCAAGAGATCCGTTGTTGAAAGTTTTAACTATTAAATAGTACTCAGACGATACTAACATTCAGCGTTTAAATATCCTCTGGCGCGCGCACCGGGCAGTACCCAGGTACGCATAGCGTACGGCGCGCCAAAGCAACAGTTTATAATGATACACAAGGGTGGGAGATCTACCCCGGGGGGCATGAGCTCTGTAATGATCCTTCCGCAGGTTCACCTACGGAAACCTTGTTACGACTTTTACTTCCTCTAAATGACCAAGACAATTACCATAACTGTGTCGTGTATATCTCCGGTTCCC
+
CCCCCGGGGGGGGGGGGGGGGGGGFGGGGGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFFEFGFFGFGGGGDFGFGGGGGGGGGGGGGGGGGGGGGEFEEFGEGGGGGGGGGFCFGGGGGGGGGGGGGGGGGGGGFFGGGGGDFGGGGGGGFECGFGDGC?FGGGGDCEC>7EED5FGGGECCEDCFGCGGGGGGGGGGGDD=G<C7+>7FCD?DFFF?FG?FF><DF5<CAF7=FFF=<4)4)7=42>BFB?B??(4<4)69<65:A<<4)),3(81844/*7*))1))3+
@M01498:244:000000000-ANT97:1:211

Seems like a mixed bag, some good reads, some bad. 

The issue is that graham and I probably need to branch immediately  into our own set of files... this could get confusing for us to keep the same document but with many parallel, different files. The  best solution may be for graham to start his own jupyter notebook, and we can each grab what we need from the other's files, especially if we stay in the same GitHub repository.

For now, since it's just me, I'll treat this document as my own dry lab notebook and keep moving...


# Creating combined fastq files

For your "average" illumina miseq results, the following section won't be useful. Graham and I, due to a unique shared run, had our samples demultiplexed for us upon arrival. To use most of the available software written for processing Illumina raw reads, however, it makes sense for us to "remultiplex" our samples into our respective studies. So if anyone in the lab is following this for their own projects, you'll probably need to skip down to the "make toy fastq files" section below. 

Right now, it seems to me that the best plan is to separate out the files that I need for a biom table, and combine them into two master forward and reverse files. These can then be paired, quality filtered, etc etc. We will need to be very careful that order is kept and all file sizes are paired. First, let's separate out just the files I want to work with.

In [68]:
## make a list for reference. Using grep to search out just "w" samples:
ls $rdir | grep -E '\-[0-9]{1,3}w_' > w-illum.txt

In [69]:
## check this:
head w-illum.txt; tail w-illum.txt

lane1-s164-index-CTAGGTGA-GACACAGTNNNN-1w_S164_L001_R1_001.fastq.gz
lane1-s164-index-CTAGGTGA-GACACAGTNNNN-1w_S164_L001_R2_001.fastq.gz
lane1-s165-index-ACGACTTG-GACACAGTNNNN-2w_S165_L001_R1_001.fastq.gz
lane1-s165-index-ACGACTTG-GACACAGTNNNN-2w_S165_L001_R2_001.fastq.gz
lane1-s166-index-GCATACAG-GACACAGTNNNN-3w_S166_L001_R1_001.fastq.gz
lane1-s166-index-GCATACAG-GACACAGTNNNN-3w_S166_L001_R2_001.fastq.gz
lane1-s167-index-TGCGAACT-GACACAGTNNNN-4w_S167_L001_R1_001.fastq.gz
lane1-s167-index-TGCGAACT-GACACAGTNNNN-4w_S167_L001_R2_001.fastq.gz
lane1-s168-index-TCTCCGAT-GACACAGTNNNN-5w_S168_L001_R1_001.fastq.gz
lane1-s168-index-TCTCCGAT-GACACAGTNNNN-5w_S168_L001_R2_001.fastq.gz
lane1-s250-index-ACCATCCA-CGCTACATNNNN-128w_S250_L001_R1_001.fastq.gz
lane1-s250-index-ACCATCCA-CGCTACATNNNN-128w_S250_L001_R2_001.fastq.gz
lane1-s251-index-CAACACCT-CGCTACATNNNN-129w_S251_L001_R1_001.fastq.gz
lane1-s251-index-CAACACCT-CGCTACATNNNN-129w_S251_L001_R2_001.fastq.gz
lane1-s252-index-CTAGGTGA-CCTAAGTCNNNN-1

In [71]:
## how many files are there with "w" in the sample name?
wc -l w-illum.txt

182 w-illum.txt


182/2 = 91 samples. That's right, a little problematic, because I had intended to exclude a few of these samples from the illumina run, as they did not show banding on a gel. Forgot to remove these.  But deal with that later.

Copy just my files into a new directory:

In [ ]:
mkdir /home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls/woodonly ## new directory

In [7]:
## save this path to a variable
rdir='/home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls/woodonly'

In [15]:
cp /home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls/*w_* $rdir ## just my wood samples

Add in positive and negative controls from both graham and I's samples, it will be informative to compare these.

In [13]:
cp /home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls/*Pos* $rdir  ## positive controls
cp /home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls/*Neg* $rdir  ## negative controls
cp /home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls/*X* $rdir ## pure culture Xylaria hypoxylon control

In [17]:
cd $rdir
ls | wc -l

196


Seems right: (91 samples + 7 controls)x2 = 196

We need to extract and combine these files, but make sure that no information is lost. 

In [18]:
gunzip $rdir/*

These will be concatenated into one large forward and one large reverse file...

A little nerve-wracking, but I think this is as simple as concatenating all R1 files and all R2 files:

In [20]:
cat $rdir/*R1* > $wdir/woodR1.fastq
cat $rdir/*R2* > $wdir/woodR2.fastq

Okay, the above works, but I don't know that what sort of order was maintained. Does this matter?

Don't think so, as long the unique sample number in the read identifier is retained. I think the main things to check are 1) are all samples included? and 2) do forward and reverse files match exactly?

1) we can check the expansion we used to make this file:

In [23]:
ls $rdir/*R1* | wc -l; ls $rdir/*R1* | wc -l

98
98


That's what we expect, 91 samples, plus 3 controls from me and and 4 from graham. 

In [21]:
ls $rdir/*R1* | head -n 3

/home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls/woodonly/lane1-s160-index-AAGCACTG-GTGATCCANNNN-Dc-X_S160_L001_R1_001.fastq
/home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls/woodonly/lane1-s161-index-AAGCACTG-TTCGTACGNNNN-Dc-PosG_S161_L001_R1_001.fastq
/home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls/woodonly/lane1-s162-index-AAGCACTG-ATGACAGGNNNN-Dc-PosI_S162_L001_R1_001.fastq


In [22]:
ls $rdir/*R1* | tail -n 3

/home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls/woodonly/lane1-s255-index-TGCGAACT-CCTAAGTCNNNN-Neg_S255_L001_R1_001.fastq
/home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls/woodonly/lane1-s256-index-TCTCCGAT-CCTAAGTCNNNN-PosG_S256_L001_R1_001.fastq
/home/daniel/Documents/Taiwan_data/wood/illumina_reads/BaseCalls/woodonly/lane1-s257-index-GAGGACTT-CCTAAGTCNNNN-PosI_S257_L001_R1_001.fastq


Looks like even the sample order was maintained. Simplifies things.

As for concern #2 from above, let's compare just the first lines of each read (the illumina sequence identifiers)

In [27]:
cd $wdir ## make sure we're in our working directory

In [28]:
## use sed to just look at every fourth line:
sed -n '1~4p' woodR1.fastq > justlabels_R1.txt
sed -n '1~4p' woodR2.fastq > justlabels_R2.txt

In [30]:
## cut everything but the individual unique identifier from these 
## = save only the last number of the illumina info, which gives each sample a unique number

cut -d ':' -f 10 justlabels_R1.txt > numsR1.txt
cut -d ':' -f 10 justlabels_R2.txt > numsR2.txt

## check differences:
diff -c numsR1.txt  numsR2.txt

(diff program returns nothing)

In [31]:
## length of files?
wc -l woodR1.fastq

22271196 woodR1.fastq


In [32]:
wc -l woodR2.fastq

22271196 woodR2.fastq


diff reports no difference, line number is the same...I think we have cleanly constructed fastq files for forward and reverse reads of the taiwan wood samples.

#### <font color='green'>2016-05-21 Dan</font>

#### <font color='green'>kernel = BASH</font>

# Make toy fastq files

Ok, now that we have two master fastq files for the wood study, what is the plan? 

The issue with these huge fastq files is that they are slow to work with, the computer can chew on them for days because they're so damn big. It seems appropriate to unload as much of the heavy computing as possible on aciss. But to effectively run aciss, it's best to do test runs as we go,  with our own computer. For that, we'll want subsetted, "toy" fastq files to work with.

How can we get a ~random sampling of our master fastq files? In my previous pipelines, I have just used the head function to take a subset of the fastq file lines, but this is not appropriate here, because the reads from each sample are clumped together in the file, and I'll exclude entire samples if I do this here. 


Have to think about this because I'll need a toy biom file that is a representative sample that I can run stats on. 

I think this random subsetting is probably best done in python. I've put together a script for this, "subset.py", included in this repository

Subset to 1,000,000 reads:

In [33]:
./subset.py woodR1.fastq 1000000 R1subset.fastq

./subset.py woodR2.fastq 1000000 R2subset.fastq

# Trim reads and Paired-end assembly

Reads always taper off in quality. Let's look at ours and decide appropriate lengths to trim to...

In [34]:
## use fastx toolkit here:
fastx_quality_stats -i R1subset.fastq -o R1subset_fastx.txt
fastx_quality_stats -i R2subset.fastq -o R2subset_fastx.txt

This creates a fastx summary table of read quality at every BP location
Normally the next step is to use the fastx script "fastq_quality_boxplot_graph.sh"
but this script was creating unreadable x-axes for me, so I modified it a bit (it's just a shell script that reads some parameters into gnuplot). So I've included a slightly modified version, "dan_plot.sh" that works a little better for me. 

In [35]:
./dan_plot.sh -i R1subset_fastx.txt -o R1subset_quality.png
./dan_plot.sh -i R2subset_fastx.txt -o R2subset_quality.png

This makes plots that look like this:

forward reads (included in repository as 'R1subset_quality'):

![R1subset_quality.png](R1subset_quality.png)

reverse reads (included in repository as 'R2subset_quality'):

![R2subset_quality.png](R2subset_quality.png)

Cutting off the low quality end of reads helps assure the accuracy of the pairing process. The question is where to cut off our reads, exactly.

It looks to me like a good cutoff would be somewhere around 255 BP for R1 and 210 for R2.

We can do this with fastx tools:

In [43]:
fastx_trimmer -l 255 -i R1subset.fastq -o R1subset_short.fastq 
fastx_trimmer -l 210 -i R2subset.fastq -o R2subset_short.fastq

and rerun the above visualizing pipeline, just to see what this looks like...

In [45]:
fastx_quality_stats -i R1subset_short.fastq -o R1subset_short_fastx.txt
fastx_quality_stats -i R2subset_short.fastq -o R2subset_short_fastx.txt

./dan_plot.sh -i R1subset_short_fastx.txt -o R1subset_short_quality.png
./dan_plot.sh -i R2subset_short_fastx.txt -o R2subset_short_quality.png

Forward reads, shortened:

![R1subset_short_quality.png](R1subset_short_quality.png)

Reverse, shortened:

![R2subset_short_quality.png](R2subset_short_quality.png)

Pairing ends- we'll use pandaseq. We can't use the randomly subsetted reads we created above, because the two files don't match exactly. So make new toy subsets that have order maintained to try out our paired-end assembly process. Then we'll make ACISS, the University of Oregon computing cluster, do the hard work on the full reads.

In [1]:
## our new toy subsets, to test the process, 100,000 reads:
head -n 400000 woodR1.fastq > toy_woodR1.fastq
head -n 400000 woodR2.fastq > toy_woodR2.fastq

In [3]:
fastx_trimmer -l 255 -i toy_woodR1.fastq -o toy_woodR1_short.fastq
fastx_trimmer -l 210 -i toy_woodR2.fastq -o toy_woodR2_short.fastq

In [4]:
## pair the reads!
pandaseq -f toy_woodR1_short.fastq -r  toy_woodR2_short.fastq -A ea_util -o 20 -F -g pairlog.txt -t .6 -w testpair.fastq

-A = use ea_util algorithm for scoring matches (using this for compatibility with Roo's leaf biom table, this is just what we did in our last table of leaf endophytes, that I would like to refer to in this paper)

-o = minimum overlap require to pair two sequences. Again, using 20 to match previous protocols. 

-F = make output fastq format, we're going to quality filtering downstream of this, even though some folks argue this may be inappropriate

-g = keep a log file of outputs besides fastq file

-t = quality cutoff that sequences must have to be retained in output fastq. Using 20 to be consistent with leaf biome table


Quick check, did all of our reads pair?

In [5]:
wc -l testpair.fastq

400000 testpair.fastq


Looks good, 400,000 lines / 4 = 100,000 reads

how do they look? Run fastx stats and plotting on them:

In [6]:
fastx_quality_stats -i testpair.fastq -o testpair_quality.txt
./dan_plot.sh -i testpair_quality.txt -o testpair_quality.png

Plot of my paired toy reads looks like this:

![testpair_quality.png](testpair_quality.png)

Let's use fastqc tools (not to be confused with fastx tools) to look at what's going on here...

In [9]:
fastqc testpair.fastq

In [10]:
## check out the report that fastqc generates with our paired reads, using firefox.
firefox $wdir/testpair_fastqc/fastqc_report.html

Of particular interest to us the following figure, of the abundances of different read lengths:

![sequence_length_distribution.png](sequence_length_distribution.png)

We truncated our sequences to 210 and 255 BP, with a 20 BP minimum required sequence overlap, so the largest possible sequences are 190+235+20 = 445 BP

The above shows that the vast majority of this subset of reads are under 300 BP in length. 

While the quality averages are lower in longer reads over 300 BP, let's retain them for the time being

Looks pretty good, with this toy subset. So now, can we do this on ACISS with our full readset?

## Using ACISS to pair reads

Let's log onto ACISS with SFTP and get our full readsets uploaded. Interactions with the computing cluster are through SSH (running programs, etc) and SFTP (uploading and downloading files). You switch constantly between these when using ACISS, so I keep one window of each going. 

This will only be possible from inside the university network, so a VPN is needed if working off campus. An account with aciss is required, but pretty easy to request: <http://aciss-computing.uoregon.edu/>

In [18]:
## make sure your in the working directory of your home coputer, with the files you need, then login
sftp dthomas@login.aciss.uoregon.edu

In [20]:
put woodR1.fastq
put woodR2.fastq

As of time of this writing, ACISS had versions of pandaseq and fastx toolkit available for use. If packages you need aren't on ACISS, the staff is really responsive, just email them!

We can check the available packages by logging in with SSH:

In [22]:
ssh dthomas@login.aciss.uoregon.edu
modul avail

To run a batch job (get ACISS to run non-interactively) we need to write and upload a script ("aciss_pair.sh"). This script is a BASH shell script, because the computing cluster uses linux. The commands in this script are basically the pipeline we just developed with the toy files above, but with our whole read sets. I've included this in my scripts folder in this repository. 

Back to sftp, logged on ACISS server:

In [ ]:
put aciss_pair.sh

Now that our files are on ACISS, we request a node (all or part of a processor) to execute the commands in a script. 

Back to SSH, on ACISS:

In [ ]:
qsub -q student aciss_pair.sh

Aciss will give us a job number that can be checked on using the que system. In this case, I was assigned "1065490.hn1". To check on this job:

In [ ]:
qstat 1065490.hn1

Okay, once aciss is chewing on this, we can log out and do other things while we wait. 

#### <font color='green'>2016-05-30 Dan</font>

#### <font color='green'>kernel = BASH</font>

Okay! Let's see how our pairing did. First, get it off of the ACISS storage. Making sure that we're still in the right working directory on home computer (use 'pwd' command if in doubt), log into ACISS with sftp and get the paired read files:

In [ ]:
## login
sftp dthomas@login.aciss.uoregon.edu
## get the paired read file
get wood_paired.fastq
## get our log for this pairing, and the fastx quality stats for all of our read files
## (I had aciss do all of this for us, see the ACISS script)
get woodpairlog.txt
get wood_paired_quality.txt 
get woodR1_short_quality.txt 
get woodR2_short_quality.txt
## and sign out
exit

For the moment, let's not clean up our ACISS storage space, until we can verify everything worked. (It is good practice to keep this space clean, once we're certain we've got these files stored and backed up elsewhere. )

How many of our reads paired?

In [12]:
wc -l woodR1.fastq
wc -l woodR2.fastq
wc -l wood_paired.fastq

22271196 woodR1.fastq
22271196 woodR2.fastq
22271196 wood_paired.fastq


looks like all reads paired.

What do our new paired reads look like? Is this an improvement over the unpaired reads? Visualize them with fastx plot:

In [6]:
## make plots of our full reads in both directions
./dan_plot.sh -i woodR1_short_quality.txt -o woodR1_short_quality.png
./dan_plot.sh -i woodR2_short_quality.txt -o woodR2_short_quality.png
## and one of our full paired reads
./dan_plot.sh -i wood_paired_quality.txt -o wood_paired_quality.png

Our full readset, ends paired:

![wood_paired_quality.png](wood_paired_quality.png)

compare this to...

(1) our forward reads:

![woodR1_short_quality.png](woodR1_short_quality.png)

(2) our reverse reads

![woodR2_short_quality.png](woodR2_short_quality.png)

looks like an improvement! 

#### <font color='green'>2016-05-31 Dan</font>

#### <font color='green'>kernel = BASH</font>

Let's do some more cleaning up. The 5' ends of our reads are of extremely variable quality. I don't want this to interfere with the OTU clustering and blast algorithms below, so I believe it's okay to get rid of this area of low quality and keep the region that we have more confidence in. 

It's worth noting that I log into ACISS to do all of these full readset manipulations. ACISS has an interactive mode that lets you use one of their nodes like an extra, powerful computer at your fingertips. You can do this on your own computer, but using ACISS keeps your machine stress-free to work on other things, so I use it as much as possible, either to run overnight with scripts like we did above, or run interactively with while I'm at my desk. To use ACISS interactively, log on with SSH (and SFTP in another window while you're at it!), and request an interactive session:

In [ ]:
## log onto ACISS with SSH
ssh dthomas@login.aciss.uoregon.edu

In [ ]:
## with ssh, ask for an interactive node:
qsub -q student -I

-q requests a particular type of node. Since I'm a student, some portion of nodes are set aside for us, so these "student" nodes are usually available to me without a wait. The actual type of node varies with this category.


-I requests an interactive session, which acts just like a BASH shell session on your own computer

In [ ]:
## on ACISS node (in SSH):

##load module (fastx)
module load fastx_toolkit

## trim reads with fastx
fastx_trimmer -f 24 -Q33 -i wood_paired.fastq -o wood_paired_clipped.fastq

## quality stats, so we can plot
fastx_quality_stats -Q33 -i wood_paired_clipped.fastq -o wood_paired_clipped_quality.txt

## ACISS has older version of fastx as of time of writing, so we in "-Q33" flag to 
## tell fastx that we are dealing with Sanger-type quality encoding, or "Phred+33"


Use sftp to get the newly created files off of ACISS:

In [ ]:
## in separate window, making sure that you are in your working directory in your home computer, log onto ACISS with SFTP
sftp dthomas@login.aciss.uoregon.edu

In [ ]:
## download the files onto your working directory
get wood_paired_clipped.fastq
get wood_paired_clipped_quality.txt

Log out of our node, so other folks can use it. 

In [ ]:
## in ACISS with SSH
exit

and plot these results with our own computer:

In [ ]:
./dan_plot.sh -i wood_paired_clipped_quality.txt -o wood_paired_clipped_quality.png

Our paired reads, with low quality 5' end removed. Looks like this:

![wood_paired_clipped_quality.png](wood_paired_clipped_quality.png)

In general, from here out I will be testing all of my commands with a toy sequence set made using the python script "subset.py" (in repository) on my own computer, then port the commands to ACISS to work on the full readset, either interactively or in a shell script. After this I will not show aciss login commands, loading modules, or sftp file transfers, etc, etc. Just use ACISS when you see fit. 

make a new toy read set, this time from our paired-end reads:

In [25]:
./subset.py wood_paired_clipped.fastq 1000000 wood_sub.fastq

I will show commands on the full read set as if performing them on my computer, but know that really I am testing commands on my own computer with the above toy read set, then often running the commands on the full read set on ACISS.

# A note about read directionality

Typically, somewhere near this point in the process, we would check read directions. Most software downstream of this point assumes a 5'-3' orientation. At the time of this writing, Illumina sequencers don't know which side of the original organismal DNA strands your read came from. About 50% of your R1 file and 50% of your R2 file should be from your original forward reads, and the other half in each from original complementary DNA. 

This is not always the case, and depends on what the folks at the sequencer facility did with your reads. In our case, uniform read directionality was established before we received our raw reads, so we're skipping this step. So check with your sequence facility to see what they've done to your reads, seek help elsewhere with checking read directionality if you need it. 

# Floating primers!

This is a thing with some ITS studies done with illumina: in some small fraction of the reads, unanchored ITS primer repeats appear in mysterious spots in the target sequence area. To date I've heard no good explanation for their presence, and varying opinions on their status as artifacts or as real data.  

Our primers were not included as part of the reads (our primers were actually used as part of the illumina sequencing primers to keep our samples separated from another study, long story!). So any primers we find are "floating" primer artifacts. I remove as many floating primers as I can find, as advised by Bálint et al. (2014). I do this with another python script. This script removes any read with an exact match to our primers or their reverse complements. However, this script leaves degenerate versions of these primer artifacts, which definitely occur. Someone with more time and expertise may want to improve on this situation...

Other studies using illumina data will need to cut off primers on the ends of their reads before using this script. 

First convert our reads from fastq to fasta format, most of our downstream processes work with fasta files. 

In [ ]:
fastq_to_fasta -n -Q33 -i wood_paired_clipped.fastq -o  wood_paired_clipped.fasta

use the script:

In [ ]:
./floatingprimers.py  wood_paired_clipped.fasta  wood_paired_clipped_defloat.fasta CTTGGTCATTTAGAGGAAGTAA GCTGCGTTCTTCATCGATGC

Arguments for this script: floatingprimers.py input_fasta_file output_fasta_file forward_primer reverse_primer. 

Reverse compliments are checked automatically from the forward and reverse primer sequences that are given as arguments. 

How many of these sequences had "floating" primers?

In [31]:
wc -l wood_paired_clipped_defloat.fasta

10941460 wood_paired_clipped_defloat.fasta


In [33]:
wc -l wood_paired_clipped.fasta

11135598 wood_paired_clipped.fasta


 11135598 / 2 - 10941460 / 2 =  (5,567,799 reads - 5,470,730 reads) = 97,069 reads 

at least 97,069 out of 5,567,799 (1.7%) of our reads contained floating primers. This phenomenon is a bit mysterious.

# Chimera checking

At this point we check our sequences to look for any sequences that may have been accidentally created from two parent organisms (an artifact of sequencing). We can do this with [usearch](http://www.drive5.com/usearch/), and a reference database of fungal sequences from [UNITE](https://unite.ut.ee/index.php). More on UNITE later...

Other folks might start to use [qiime](http://qiime.org/) at this point in the process, with the command [identify_chimeric_seqs.py](http://qiime.org/tutorials/chimera_checking.html). I do not use qiime here, can't seem to make it work. Their "chimera slayer" option requires aligned reads, which is pointless with ITS, and their usearch option is out of date, behaves weirdly. 

So we'll use [usearch](http://www.drive5.com/usearch/) directly. This is the default algorithm behind a lot of qiime's scripts anyway, and this will enable us to use the latest versions of usearch, which qiime doesn't seem to have incorporated, anyway. When you download usearch, it comes as a small binary file, you'll want to rename it to something handy and put it in your working directory, on your computer and on ACISS. At the time of this writing, I downloaded usearch v8.1.1861, this was the latest and has a lot of downstream capabilities that were previously lacking, like the ability to create biome tables. 

usearch prefers to have a high-quality reference database to check for parent sequences that may have created daughter chimera reads. We'll download UNITE database for this. They have a version just for this family of algorithms:

In [ ]:
## command line download of the reference set. (Can also go to the website with a browser and click on the download)
wget https://unite.ut.ee/sh_files/uchime_reference_dataset_01.01.2016.zip

In [ ]:
## unpackage it
unzip uchime_reference_dataset_01.01.2016.zip

This directory contains several versions of UNITE. Of particular interest to us is an ITS1 only database, since our reads are from the ITS1 region only. We'll use this as our reference database with usearch:

In [45]:
## use uclust algorithm directly to check for chimeras for us. 

usearch -uchime_ref wood_paired_clipped_defloat.fasta -db ~/uchime_reference_dataset_01.01.2016/ITS1_ITS2_datasets/uchime_sh_refs_dynamic_develop_985_01.01.2016.ITS1.fasta -nonchimeras wood_notchim.fasta -strand plus -uchimeout chim_log.txt

Usearch returns a FASTA file of sequences that are NOT chimeras, let's see how many reads were lost in this process.

Usearch spits out multi-line fastas, meaning they insert line breaks inside of the sequences. Don't really know why. But to count reads in these we need to grep the sequence identifier line, instead of just using the 'wc -l' command:

In [46]:
## grep the beginning of each read:
grep '\<M' wood_notchim.fasta | wc -l
grep '\<M' wood_paired_clipped_defloat.fasta | wc -l

5423495
5470730


5,470,730 reads before chimer check - 5,423,495 reads after chimera check = 47,235 chimera sequences! Seems like a lot. 


So this works, but usearch likes to return a multiline fasta, with line breaks after a certain number of BP. We may need to fix this...

<a id='otus'></a>

# Pick otus and representative sequences:

Qiime has a pretty classic command for clustering otus, [pick_otus.py](http://qiime.org/scripts/pick_otus.html). 

However, as always, Qiime and I are not getting along. The aciss installation of qiime seems unreliable at this time. So we'll use usearch, the default algorithm used by qiime for out clustering, just without the qiime wrapper. 

We'll use the 95% cutoff, to keep things consistent with Roo's biom table. The justification is that the ITS 1 region is really variable, and more wobble is necessary among "species" concepts here to recapture the standard 97% or so cutoff of full ITS comparisons. 

We start by "dereplicating" our sequencings, which reduces our read file to unique sequences (no repeats), and a size tag to indicate how many sequences were originally present. I think. Not sure, in the past, qiime took care of this for me. 

In aciss, load the latest usearch. ACISS has the license to 64-bit usearch algorithms, which is handy for these large illumina data sets:

In [ ]:
module load usearch/8.1.1803

We start by "dereplicating" our sequencings, which reduces our read file to unique sequences (no repeats), and a size tag to indicate how many sequences were originally present. I think. Not sure, in the past, qiime took care of this for me. 

In [ ]:
usearch -derep_fulllength wood_notchim.fasta -fastaout wood_notchim_derep.fasta -sizeout

the "-sizeout" maintains a record of how sequences are represented exactly (100% similarity) by each read in the dereplicated file. 

The we sort our files by this number:

In [ ]:
usearch -sortbysize wood_notchim_derep.fasta -fastaout wood_notchim_sorted.fasta

And cluster them. Usearch has several methods for clustering. At the time of writing, the latest, preferred algorithm is [-cluster_otus](http://drive5.com/usearch/manual/cmd_cluster_otus.html). But this algorithm has only been tested for the .97 similarity radius for otus, and the author, Robert Edgar, will not vouch for it at any other radius. So let's try a [different algorithm](http://drive5.com/usearch/manual/cmd_cluster_smallmem.html), which I believe is the descendent of the uclust algorithm that Qiime uses in its scripts, anyway:

In [1]:
usearch -cluster_smallmem wood_notchim_sorted.fasta -id 0.95 -centroids otus_95_uclust.fasta -sizein -sizeout -sortedby size 

We keep the size ("-sizeout"), this might be handy information. 

In the same ACISS script I will also ask ACISS to do a .97 similarity OTU fasta file while I'm at it. This may come in handy for comparisons, or may be preferable (need to do more searching in the literature to see what other folks are up to with ITS1 studies). This script is in the repository ACISS scripts folder. This took ~40 hours for each clustering event, so be patient, or find a more efficient algorithm!

Also, note that when the above command has finished, and otu clusters are ready, this point or after [assigning taxonomy](#taxonomy) may be a good spot to clean up our otu labels, though I put this off till [later](#sed2).

On that note, I will also check another algorithm here, an alternative to usearch/uclust, called [swarm](https://github.com/torognes/swarm). I don't think this one is more time-efficient, but it is an open-source alternative to the usearch option, and some claim it is superior to recapturing mock community structure. I will run this at the same otu dissimilarity radius (.95 and .97). If there's time, we'll check all of these cluster sets to see which recovers our mock communities most accurately (if the mock community samples are intact, we can't check this yet!). 

So swarm:

In [ ]:
## or not...waiting on some information about how to use swarm...

# Customize UNITE database

UNITE is an attempt by mycologists to get some quality control on fungal accessions in large public databases like GenBank. We will use it to identify our sequences. 

It's handy, but last I checked, there were some problems with sequences only identified to very low taxonomic resolution, e. g. accessions identified only to phylum. The cause problems because BLAST may display these matches over other very close matches that are more completely identified, if they match ever-so-slightly-better to the poorly identified sequences.

I will use usearch's utax algorithm. So let's get the UNITE database made expressly for this:

In [ ]:
wget https://unite.ut.ee/sh_files/utax_reference_dataset_31.01.2016.zip


I had lot of trouble getting my python interpreter to work on the database due issues with special characters like umlauts and other symbols (especially "ë" and "×" that are used in some names of fungi in UNITE. My versions of sed and iconv and my python interpeter would not work with UNITE names. My $LANG = en_US.UTF-8, which seems like it should be able to handle these characters, but character encoding is a complex subject which I do not understand. Most of the database is in simple unicode8, though, so I was able to use my favorite text editor [VIM](www.vim.org) to substitute these characters out:

In [ ]:
unzip utax_reference_dataset_31.01.2016.zip
vim utax_reference_dataset_31.01.2016.fasta

Inside vim command mode (press ':'):

In [ ]:
:%s/ë/e/g
:%s/×/x/g

This fixed it for me. Good luck with your unicode. Maybe mac users won't have this issue. Exit vim, back in the shell. Here I've written another python script, "UNITE_with_class2.py", for only keeping those UNITE records with class identification or better. This is in the repository. 

In [ ]:
./UNITE_with_class2.py utax_reference_dataset_31.01.2016.fasta utax_ref_class.fasta

And convert this "classy" database to a format that usearch/utax likes:

In [ ]:
usearch -makeudb_utax utax_ref_class.fasta -output utax_ref_class.udb -report dbmake_report.txt

<a id='taxonomy'></a>

# Assign taxonomy to otus

Time to put some names on the otu reference set. To do this with the usearch pipeline, we need to give all of our otu clusters a unique name/number. Since we used [usearch -cluster_smallmem](http://www.drive5.com/usearch/manual/cmd_cluster_smallmem.html) instead of [usearch -cluster_otus](http://www.drive5.com/usearch/manual/cmd_cluster_otus.html) due to our .95 similarity radius for our otus, [see here for discussion](http://www.drive5.com/usearch/manual/uparse_otu_radius.html), we need to add our otu names ourselves. As everyone knows, I like to do things the hard way. Anyway, wrote a python script for this, called "addOTUtag.py", it's in the repository. 

Also, note that here or in [otu clustering](#otus) may be a good spot to clean up our otu labels, though I put this off till [later](#sed2).

In [8]:
./addOTUtag.py otus_95_uclust.fasta OTU otus_95_uclust_relabel.fasta

Three arguments: old fasta file, label ("OTU" recommended), new fasta file. 

Let's try to assign taxonomy to the .95 radius otu clusters we just generated...

In [9]:
 usearch81 -utax otus_95_uclust_relabel.fasta -db utax_ref_class.udb -strand both -fastaout otus_95_uclust_ass_tax.fasta

usearch v8.1.1861_i86linux32, 4.0Gb RAM (8.1Gb total), 8 cores
(C) Copyright 2013-15 Robert C. Edgar, all rights reserved.
http://drive5.com/usearch

Licensed to: dthomas@uoregon.edu


00:01 117Mb Read taxonomy info...done.
00:01 119Mb Reading pointers...done.
00:02 124Mb Reading db seqs...done.



I should note that I had trouble creating the reference database above with the 64 bit version of usearch on aciss, so I ran the above commands with the usearch 32 bit version on my own computer. The database creation took about 4 hours to run on my machine, but the taxonomic assignments were quick, ~40 min. 

Important to note that two steps above, making the utax data database and assigning taxonomy, need to be done with the same version of usearch, either the 32 or 64 bit version. Using one version for one step and the other version for the next does not work. For example, don't create your database at home with the 32-bit version, then try to have aciss use the 64-bit version use this database to assign taxonomy. Won't work. 

# Make biome table

Since I've been using USEARCH for the pipeline so far, let's see if it can take us all the way. To make our biome table, usearch needs clear [sample identifiers](http://www.drive5.com/usearch/manual/upp_labels_sample.html) to discern from which sample our reads originated. USEARCH looks for "sample=" to identify sample number. If it doesn't find it, it takes the first sequence of letters and numbers in our sequence indentifiers, which won't work for us. 

Graham's and my read identifiers look like this:

In [11]:
head -n 1 wood_notchim.fasta

>M01498:244:000000000-ANT97:1:1101:7581:1916:160


The information about this sequencing run is not really important for us. And the unique identifiers for each read (the second-to-last three sets of numbers) are probably not useful downstream of this point either. The good stuff is the last three digits, which tell which of my samples this read came from. So let's get rid of all the rest with [sed](https://www.gnu.org/software/sed/):

<a id='sed'></a>

In [ ]:
sed s/'M.*:[0-9]:\([0-9]\{3,\}:\)\{3\}'/sample=/g  wood_notchim.fasta > wood_notchim_cleanlab.fasta

sed is a streaming text editor that uses [regular expression syntax](http://www.regular-expressions.info/). So the above 
command looks complex. If you are using this as a guide for your own reads, you may have to tweak the above command or find another way to modify your labels, they're going to be different than mine. The resulting simplified read identifier line: 

In [13]:
head -n 1 wood_notchim_cleanlab.fasta

>sample=160


Using the usearch [-usearch_global](http://drive5.com/usearch/manual/cmd_usearch_global.html) program. We use our OTUs with assigned taxonomy as a reference database:

In [ ]:
usearch -usearch_global wood_notchim_cleanlab.fasta -db otus_95_uclust_ass_tax.fasta -strand both -id 0.95 -biomout wood_otu_uclust_95.biom

This only worked for me using the more powerful 64-bit version of usearch. The "-id" argument I lined up with otu similarity cutoff from our otu clustering above, 95% similarity. 

## Side trip: Make toy biome table:

The resulting biome table is very large, difficult to check for accuracy, so I'll create a toy version. First I need script to subset randomly from fasta files (the previous subsetting was done from fastq files.). I also need to finally make a script that converts multine fastas to two-line fastas. These are in the repository. 

In [14]:
./fasta_remove_linebreaks.py wood_notchim.fasta wood_notchim_nolb.fasta

In [15]:
./subset_fasta.py wood_notchim_nolb.fasta 1000000 wood_notchim_nolb_sub.fasta

In [16]:
sed s/'M.*:[0-9]:\([0-9]\{3,\}:\)\{3\}'/sample=/g  wood_notchim_nolb_sub.fasta > wood_notchim_cleanlab_sub.fasta

On ACISS:

In [ ]:
usearch -usearch_global wood_notchim_cleanlab_sub.fasta -db otus_95_uclust_ass_tax.fasta -strand both -id 0.95 -biomout otu_95_sub.biom

# Checking and tidying our biome tables

We've got our biome tables! As of the time of this writing, usearch returns biome tables in the [JSON format](http://biom-format.org/documentation/format_versions/biom-1.0.html). This is the older of two possible formats for biome tables, and we may need to convert to the newer [HDF5 format](http://biom-format.org/documentation/format_versions/biom-2.0.html). But for now we'll keep this format and take a look, it's a pretty human readable format.

In [2]:
head -n 12 wood_otu_uclust_95.biom

{
	"id":"wood_otu_uclust_95.biom",
	"format": "Biological Observation Matrix 1.0",
	"format_url": "http://biom-format.org",
	"generated_by": "usearch8.11803",
	"type": "OTU table",
	"date": "Thu Jun 23 03:45:12 2016",
	"matrix_type": "sparse",
	"matrix_element_type": "float",
	"shape": [4009353,98],
	"rows":[
		{"id":"OTU1070430:M01498:244:000000000-ANT97:1:1101:5958:4272:160", "metadata":{"taxonomy":"d:Fungi(1.0000),p:Ascomycota(0.9036),c:Sordariomycetes(0.7215),o:Xylariales(0.6231),f:Xylariaceae(0.4147),g:Xylaria(0.1335),s:Xylaria_hypoxylon_SH205537.07FU"}},


This is confusing to look at, but it's pretty simple. The "shape" attribute tells us we have a matrix with 4009353 rows and 98 columns. The number of columns should be equal to the number of samples, which is correct here. The number of rows should be roughly equal to the number of OTU's in our OTU file:

In [4]:
grep '>' otus_95_uclust_ass_tax.fasta | wc -l

4019831


Looks okay. Some otus apparently did not get used, I am a little confused by this, since all OTUs were originally a read. Most of these are singletons and will eventually be thrown out, in downstream ecological analyses, so hopefully this doesn't matter too much. We'll definitely keep our OTU file with all the original read info for future reference. 

Below this is an attribute call "rows". This tells us what our otus are called, and what metadata is associated with them. It's a mess!:

In [7]:
grep "\<rows" wood_otu_uclust_95.biom -A 3

	"rows":[
		{"id":"OTU1070430:M01498:244:000000000-ANT97:1:1101:5958:4272:160", "metadata":{"taxonomy":"d:Fungi(1.0000),p:Ascomycota(0.9036),c:Sordariomycetes(0.7215),o:Xylariales(0.6231),f:Xylariaceae(0.4147),g:Xylaria(0.1335),s:Xylaria_hypoxylon_SH205537.07FU"}},
		{"id":"OTU2732738:M01498:244:000000000-ANT97:1:1101:5733:4572:160", "metadata":{"taxonomy":"d:Fungi(0.9999),p:Ascomycota(0.6978),c:Sordariomycetes(0.5003),o:Xylariales(0.3662),f:Xylariaceae(0.1857),g:Xylaria(0.0446),s:Xylaria_hypoxylon_SH205537.07FU"}},
		{"id":"OTU2406308:M01498:244:000000000-ANT97:1:1101:11114:2667:160", "metadata":{"taxonomy":"d:Fungi(0.9994),p:Ascomycota(0.5394),c:Sordariomycetes(0.2329),o:Xylariales(0.1266),f:Xylariaceae(0.0793),g:Xylaria(0.0105),s:Xylaria_papulis_SH199084.07FU"}},


The only important info here is our otu number and our taxonomy metadata. To prevent this mess I could have been more careful above, when crafting our [OTU reference data base](#otus) or [assigning taxonomy](#taxonomy). But for now, I'll clean this up with sed again, similar to what we did when [simplifying our read names](#sed) for making the biom table.

<a id='sed2'></a>

In [ ]:
sed s/':M.*:[0-9]:\([0-9]\{3,\}:\)\{3\}[0-9]\{3\}'//g wood_otu_uclust_95.biom > wood_otu_uclust_95_relab.biom

This looks complicated, but I am using sed to remove the parts of my row labels that I don't want. The row labels only retain the otu number and taxonomy metadata. They look like this:

In [11]:
grep "\<rows" wood_otu_uclust_95_relab.biom -A 1

	"rows":[
		{"id":"OTU1070430", "metadata":{"taxonomy":"d:Fungi(1.0000),p:Ascomycota(0.9036),c:Sordariomycetes(0.7215),o:Xylariales(0.6231),f:Xylariaceae(0.4147),g:Xylaria(0.1335),s:Xylaria_hypoxylon_SH205537.07FU"}},


The other problem here is that our taxonomy assignments are really messy. Downstream applications like biom, phyloseq, and phinch can't read this biom table yet. This may be due to the format of our taxonomy data, though the only good example of taxonomy metadata formatting I can find is [here](http://biom-format.org/documentation/adding_metadata.html). To fix this with sed is asking a lot, so I'll write a script for this in python. 

In [24]:
./format_tax.py wood_otu_uclust_95_relab.biom wood_otu_uclust_95_relab2.biom

Row metadata now looks like this:

In [25]:
grep "\<rows" wood_otu_uclust_95_relab2.biom -A 1

	"rows":[
		{"id":"OTU1070430", "metadata":{"taxonomy": "Root;k__Fungi,p__Ascomycota,c__Sordariomycetes,o__Xylariales,f__Xylariaceae,g__Xylaria,s__Xylaria_hypoxylon_SH205537.07FU"}},


We can do an initial check to see how successful our biom table formation is by using the validate-table command:

<a id='sam_meta'></a>

# Adding sample metadata

There are two types of metadata, one for our otus ("observation" or row metadata) that is typically taxonomy. The other is information about our samples (column metadata). The taxonomy is added [above](#taxonomy). Sample metadata is added using the set of python scripts provided by the folks behind the [BIOM project](http://biom-format.org/index.html). These sample metadata are what differentiate our samples and enable us to ask ecological questions about our data.

What do our columns headers look like right now, without metadata?:

In [14]:
grep "\<columns" wood_otu_uclust_95_relab.biom -A 3

	"columns":[
		{"id":"160", "metadata":null},
		{"id":"161", "metadata":null},
		{"id":"162", "metadata":null},


Obviously, sample metadata will vary heavily between studies, because this is where information about treatments (in experimental studies) or information about locations and ecological conditions of samples (in observational studies) is stored. For now, I will create a sample metadata that simply maps these samples back to the names I gave them originally. Eventually, I will create a more detailed table with X and Y coordinates, and environmental variable of each site. 

### Making a simple sample metadata table from my file names

What follows is a pipe for making a metadata table from my original raw read file names. This is going to be pretty unique to graham's and my data. Might be handy to look at how this was done, but will definitely not be a cut-and-paste section for other read sets. 

Going back to our raw reads, the file names serve as a reference for mapping the numbers given by illumina to our samples (what is currently our column names in our biom table) to the original names that we gave our samples. For example, Graham's single species control "Dc-X" was assigned sample number 160 by the illumina sequencer software:

In [26]:
ls $rdir | head -n 12

lane1-s160-index-AAGCACTG-GTGATCCANNNN-Dc-X_S160_L001_R1_001.fastq
lane1-s160-index-AAGCACTG-GTGATCCANNNN-Dc-X_S160_L001_R2_001.fastq
lane1-s161-index-AAGCACTG-TTCGTACGNNNN-Dc-PosG_S161_L001_R1_001.fastq
lane1-s161-index-AAGCACTG-TTCGTACGNNNN-Dc-PosG_S161_L001_R2_001.fastq
lane1-s162-index-AAGCACTG-ATGACAGGNNNN-Dc-PosI_S162_L001_R1_001.fastq
lane1-s162-index-AAGCACTG-ATGACAGGNNNN-Dc-PosI_S162_L001_R2_001.fastq
lane1-s163-index-AAGCACTG-CGACCTAANNNN-Dc-Neg_S163_L001_R1_001.fastq
lane1-s163-index-AAGCACTG-CGACCTAANNNN-Dc-Neg_S163_L001_R2_001.fastq
lane1-s164-index-CTAGGTGA-GACACAGTNNNN-1w_S164_L001_R1_001.fastq
lane1-s164-index-CTAGGTGA-GACACAGTNNNN-1w_S164_L001_R2_001.fastq
lane1-s165-index-ACGACTTG-GACACAGTNNNN-2w_S165_L001_R1_001.fastq
lane1-s165-index-ACGACTTG-GACACAGTNNNN-2w_S165_L001_R2_001.fastq


Let's save these file names as a text file:

In [27]:
ls > $rdir/filenames.txt

Some cutting with sed, to get just sample name and sample number, with a tab seperator:

In [13]:
sed s/'.*N\{4\}-'//g filenames.txt | sed s/_L001.*//g | sed s/_S/\\\t/g > sample_metadata.tsv

Take the remaining two columns of text, rearrange them: 

In [14]:
paste <(cut -f2 sample_metadata.tsv) <(cut -f1 sample_metadata.tsv) > sample_metadata_head.tsv

Insert a header with sed:

In [15]:
sed -i '1i#SampleID\tSampleName' sample_metadata_head.tsv

In [16]:
mv sample_metadata_head.tsv sample_metadata.tsv

We have doubled entries, because this came from our raw reads, which have forward and reverse files. Get rid of these with uniq. And take a look at it. 

In [22]:
uniq sample_metadata.tsv > sample_metadata_U.tsv
mv sample_metadata_U.tsv sample_metadata.tsv
head sample_metadata.tsv

#SampleID	SampleName
160	Dc-X
161	Dc-PosG
162	Dc-PosI
163	Dc-Neg
164	1w
165	2w
166	3w
167	4w
168	5w


Looks like our first sample metadata. Let's try to attach it.

### Adding metadata

In [ ]:
biom add-metadata -i otu_95_sub_relab3.biom -o otu_95_sub_wMeta.biom -m sample_metadata.tsv --output-as-json